### ISSUE TITLE
Support Dependencies

### ISSUE THREAD

**Issue by kylebrandt** [bosun-monitor/bosun#431]
Problem: Something goes down which results in lots of other things being down, because of this, we get a lot of alerts.

Common Examples:
 * A Network Partition: Some portion of hosts become unavailable from  bosun's perspective
 * Host Goes Down: Everything monitored on that host becomes unavailable
 * Service dependencies: We expect some service to go down if another service goes down
 * Bosun can't query it's database (This is probably a different feature, but noting here nonetheless)

Things I want to be able to do based on our config at Stack Exchange:
 * Have our host-based alert macro include detect if the host in Oregon (because the host name has "or-". So this is basically a dependency based on a lookup table
 * Have our host-based alerts not trigger if bosun is unable to ping the host (which would be another alert most likely)
 * Be able to have dependencies for alerts that may have no group. 

The status for any alert that is not triggering for an alert should be "unevaluated". This won't show up on the dashboard or trigger notifications.

Two general approaches come to mind. The first is that dependencies require another alert. That other alert is run first, and the alert won't trigger based on the result of another alert. The other is that dependencies are an expression. I think the expression route only really makes sense if an alert itself can be used as an expression.

Another possibility which I haven't thought much about is that alerts generate dependencies and not the other way around. So for example, an alert marks some tagset as something that should not be evaluated. 

Making Stuff Up....

```
macro ping_location {
    template = ping.location
    $pq = max(q("sum:bosun.ping.timeout{dst_host=$loc*,host=$source}", "5m", ""))
    $grouped = t($pq,"")
    $hosts_timing_out = sum($grouped)
    $total_hosts = len($grouped)
    $percent_timeout = $hosts_timing_out / $total_hosts * 100
    crit = $percent_timeout > 10
}

#group is empty
alert or_hosts_down {
    $source=ny-bosun01
    $loc = or-
    $name = OR Peak
    macro = ping_location
}

#Group is {dst_host=*}
alert host_down {
   template = host_down
   cirt = max(q("sum:bosun.ping.timeout{dst_host=*", "5m", ""))
}

lookup location {
    entry host=or-* {
        alert = alert("or_hosts_down")
    }
    ...
}

macro host_based {
   #This makes it so alerts based on this macro that are host based won't trigger if 
   dependency = lookup("location", "alert") || alert("host_down")
   #Another idea here is that you can create tag synonyms for an alert. So instead of having to add this lookup function that translates, have a synonym feature of alerts and also global that says (consider this tag key to be the same as this tag key). This would also solve an issue with silences (i.e. silencing host=ny-web11 doesn't do anything for the haproxy alert that has hosts as svname). Another issue with that is the those alerts are not tag based, so we actually need inhibit in that case. 
}

```

**Comment by Dieterbe** [comment]
> The other is that dependencies are an expression. I think the expression route only really makes
> sense if an alert itself can be used as an expression.

would it be useful if inside an alert you could reference `<other_alert>.crit` and "pull in" that logic (maybe with cached values)?

**Comment by Dieterbe** [comment]
so seeing #723 .. does that mean you SE folks figured out how you want dependencies to be handled? mind sharing the ideas?

**Comment by kylebrandt** [comment]
Going with what Matt has under "feature is". What open questions do you
have?

**Comment by mjibson** [comment]
Current thoughts are:

1. new `depends` directive in the `alert` section which is an expression
2. during the schedule run, depends is run. for any alert with the same tags as the depends whose value is != 0, that alert's state is changed to `unevaluated`

Open questions:

1. where does the unevaluated state fit in? between normal and warning? need to do a bit of auditing to make sure it won't disrupt too much stuff. for example, the javascript needs to learn about it.
2. how do we want to handle state changes to/from unevaluated? crit -> unevaluated does what? we'll probably come up with a simple rule like "treat it exactly like normal", but just need to think through consequences.

Future features:

1. add a new `alert` function to the expression language that takes the name of an alert and `crit` or `warn` strings which returns the result of that alert's evaluation. this one is complicated because expressions could return a series, so what is the return type of this function? just number? probably good enough for now. but what do we do when it's a scalar? lots of annoying things here.

**Comment by kylebrandt** [comment]
What if unevaluated was not a state. Rather it acts as an additional flag that has the following effects:

-. Notifications are not sent for anything with unevaluated set
- No State is changed
- Existing active alerts get a icon that indicates they are currently unevaluated
- Other stuff that is unevaluated is not displayed anywhere (except maybe an api route just so it is accessible)
- When the unevaluated flag is cleared, checks resume and state is updated.

Basically the idea is since it is unevaluated, there is no state change, because we are not evaluating it.

**Comment by kylebrandt** [comment]
![image](https://cloud.githubusercontent.com/assets/1692624/6221840/58f655ac-b617-11e4-8080-ec28951d61e4.png)

**Comment by Dieterbe** [comment]
to be more complete:
![chat](https://cloud.githubusercontent.com/assets/20774/6221876/3132ddd2-b618-11e4-85a3-7d4b0fd687bd.png)

**Comment by Dieterbe** [comment]
> add a new alert function to the expression language that takes the name of an alert and crit or warn strings which returns the result of that alert's evaluation. this one is complicated because expressions could return a series, so what is the return type of this function?

aren't series illegal for warn/crit expressions? i thought they only could (or at least should) be scalars or numbers.

> just number? probably good enough for now. but what do we do when it's a scalar? lots of annoying things here.

could we just define an interface type and both scalar and number could implement the interface. and that's what the alert function could return.
or how about an `alert` directive doesn't actually invoke a function, but rather it expands in place the text contents (expression) of the specified "remote" expression? that way it would be as if the user had put the exact same expression straight into the depend directive.

**Comment by kylebrandt** [comment]
WIP https://github.com/bosun-monitor/bosun/pull/723

**Comment by kylebrandt** [comment]
Case Normal: 
   return 0
Case Warn:
   return warn_value
Case Crit:
   return crit_value
Default (error, unknown, unevaluated):
	return 1

Depends still an expression. We are adding the ability reference other alerts. We are doing this via function, that has certain return values as shown above. We are doing this because we want only the highest level dependency even if that dependency is ownknown. 

We are also adding the ability to remove tags and this is so dependencies don't need a different type of join logic and can continue to use subset logic.

Dependencies need to be resolved at config time, so you no the order to run alerts in.

**Comment by kylebrandt** [comment]
Actually instead of returning the crit/warn values from alert, maybe we should use constants to represent the states. That way I can distinguish between Crit/Wan inside other alerts that call alert().

**Comment by kylebrandt** [comment]
A downside with this method is that you won't have access to any data inside another alert. So for example if call Alert("foo") in alert("baz") - We won't have access to details of Alert("baz") in Foo's notification template.

**Comment by gbrayut** [comment]
This was added in March, but currently is [not documented](https://github.com/bosun-monitor/bosun-monitor.github.io/search?utf8=%E2%9C%93&q=depends). Adding Label for Needs Documentation.

**Comment by nvhoang** [comment]
This is a very useful feature. I am glad that you guys already looked at it. I tried it today and got a question. Why is that the alert and its dependent alert needs to share at least one tag?

In my use case, my main alert is triggered if the latency of pipeline A is high. But if there's another pipeline B is already running, the latency rise of A is expected and the alert should be ignored.

**Comment by kylebrandt** [comment]
See the ungroup expression func

**Comment by captncraig** [comment]
@kylebrandt I think the problem is that `empty` and `empty` are not valid tagsets for dependencies. I think maybe they should be.

**Comment by kylebrandt** [comment]
If that is the case I would like to add back regroup

**Comment by nvhoang** [comment]
Craig was right. Empty sets are not valid for dependency.

What's regroup? Is it for assigning a group to a series?

Also, I haven't tried but is it possible to have multiple "depends" for an
alert?
On Jun 4, 2015 10:34 PM, "Kyle Brandt" <notifications@github.com> wrote:
>
> If that is the case I would like to add back regroup
> On Jun 4, 2015 11:33 AM, "Craig Peterson" <notifications@github.com>
wrote:
>
> > @kylebrandt <https://github.com/kylebrandt> I think the problem is that
>
> > empty and empty are not valid tagsets for dependencies. I think maybe
> > they should be.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
https://github.com/bosun-monitor/bosun/issues/431#issuecomment-108936772>

**Comment by kylebrandt** [comment]
Regroup is a function we used to have that allowed you to change the group.
It was removed but I hit a use case the other day. You can hack around it
with transpose, but we should just bring back that func.

You can have dependencies chains. Depends is just an expression, so you can
combine different alert() with bool IPs like && or || for example.
